### ISSUE TITLE
ValueError: Expected parameter covariance_matrix

### ISSUE THREAD

**Issue by eichlerwest** [BGU-CS-VIL/DeepDPM#44]
I see where this issue was brought up before, but it is not resolved.

Here is the error:

Traceback (most recent call last):
  File "/home/eich467/AHA-ToDo/DeepDPM/DeepDPM_alternations.py", line 236, in <module>
    train_clusternet_with_alternations()
  File "/home/eich467/AHA-ToDo/DeepDPM/DeepDPM_alternations.py", line 205, in train_clusternet_with_alternations
    trainer.fit(model, train_loader, val_loader)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 499, in fit
    self.dispatch()
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 546, in dispatch
    self.accelerator.start_training(self)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/accelerators/accelerator.py", line 73, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 114, in start_training
    self._results = trainer.run_train()
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 633, in run_train
    self.train_loop.on_train_epoch_start(epoch)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/trainer/training_loop.py", line 203, in on_train_epoch_start
    self.trainer.call_hook("on_train_epoch_start")
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1102, in call_hook
    output = hook_fx(*args, **kwargs)
  File "/home/eich467/AHA-ToDo/DeepDPM/src/AE_ClusterPipeline.py", line 251, in on_train_epoch_start
    self._init_clusters()
  File "/home/eich467/AHA-ToDo/DeepDPM/src/AE_ClusterPipeline.py", line 121, in _init_clusters
    self.clustering.init_cluster(self.train_dataloader(), self.val_dataloader(), logger=self.logger, centers=centers, init_num=self.init_clusternet_num)
  File "/home/eich467/AHA-ToDo/DeepDPM/src/clustering_models/clusternet.py", line 57, in init_cluster
    self.fit_cluster(train_loader, val_loader, logger, centers)
  File "/home/eich467/AHA-ToDo/DeepDPM/src/clustering_models/clusternet.py", line 65, in fit_cluster
    self.model.fit(train_loader, val_loader, logger, self.args.train_cluster_net, centers=centers)
  File "/home/eich467/AHA-ToDo/DeepDPM/src/clustering_models/clusternet_modules/clusternet_trainer.py", line 42, in fit
    cluster_trainer.fit(self.cluster_model, train_loader, val_loader)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 499, in fit
    self.dispatch()
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 546, in dispatch
    self.accelerator.start_training(self)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/accelerators/accelerator.py", line 73, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 114, in start_training
    self._results = trainer.run_train()
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 637, in run_train
    self.train_loop.run_training_epoch()
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/trainer/training_loop.py", line 577, in run_training_epoch
    self.trainer.run_evaluation(on_epoch=True)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 726, in run_evaluation
    output = self.evaluation_loop.evaluation_step(batch, batch_idx, dataloader_idx)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/trainer/evaluation_loop.py", line 166, in evaluation_step
    output = self.trainer.accelerator.validation_step(args)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/accelerators/accelerator.py", line 177, in validation_step
    return self.training_type_plugin.validation_step(*args)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 131, in validation_step
    return self.lightning_module.validation_step(*args, **kwargs)
  File "/home/eich467/AHA-ToDo/DeepDPM/src/clustering_models/clusternet_modules/clusternetasmodel.py", line 283, in validation_step
    cluster_loss = self.training_utils.cluster_loss_function(
  File "/home/eich467/AHA-ToDo/DeepDPM/src/clustering_models/clusternet_modules/utils/training_utils.py", line 235, in cluster_loss_function
    gmm_k = MultivariateNormal(model_mus[k].double().to(device=self.device), model_covs[k].double().to(device=self.device))
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/torch/distributions/multivariate_normal.py", line 146, in __init__
    super(MultivariateNormal, self).__init__(batch_shape, event_shape, validate_args=validate_args)
  File "/data/eich467/miniconda3/envs/DPM/lib/python3.9/site-packages/torch/distributions/distribution.py", line 55, in __init__
    raise ValueError(
ValueError: Expected parameter covariance_matrix (Tensor of shape (10, 10)) of distribution MultivariateNormal(loc: torch.Size([10]), covariance_matrix: torch.Size([10, 10])) to satisfy the constraint PositiveDefinite(), but found invalid values:
tensor([[ 0.2024,  0.0064, -0.0058, -0.0042, -0.0496,  0.0119,  0.0353,  0.0382,
         -0.0344, -0.0670],
        [ 0.0064,  0.3373,  0.0737,  0.0591, -0.0159,  0.0727, -0.1940,  0.1438,
          0.2550, -0.2739],
        [-0.0058,  0.0737,  0.2339, -0.0314,  0.0313, -0.0084, -0.0907,  0.0229,
          0.0959, -0.0558],
        [-0.0042,  0.0591, -0.0314,  0.2218,  0.0188,  0.0588, -0.0070,  0.0613,
          0.0971, -0.0632],
        [-0.0496, -0.0159,  0.0313,  0.0188,  0.6186,  0.0445, -0.0183, -0.2485,
          0.1914,  0.3667],
        [ 0.0119,  0.0727, -0.0084,  0.0588,  0.0445,  0.2089, -0.0516, -0.0080,
          0.1229, -0.0549],
        [ 0.0353, -0.1940, -0.0907, -0.0070, -0.0183, -0.0516,  0.4900, -0.0818,
         -0.1718,  0.2499],
        [ 0.0382,  0.1438,  0.0229,  0.0613, -0.2485, -0.0080, -0.0818,  0.5531,
         -0.1021, -0.2932],
        [-0.0344,  0.2550,  0.0959,  0.0971,  0.1914,  0.1229, -0.1718, -0.1021,
          0.8345, -0.2239],
        [-0.0670, -0.2739, -0.0558, -0.0632,  0.3667, -0.0549,  0.2499, -0.2932,
         -0.2239,  0.8357]], device='cuda:0', dtype=torch.float64)

Here are the arguments I used to execute the code:

python DeepDPM_alternations.py --latent_dim 10 --dataset mnist --lambda_ 0.005 --lr 0.002 --init_k 3 --train_cluster_net 200 --alternate --init_cluster_net_using_centers --reinit_net_at_alternation --dir /home/eich467/AHA-Archived/DCC/data/mnist/MNIST/raw --pretrain_path ./saved_models/ae_weights/mnist_e2e.zip --number_of_ae_alternations 3 --transform_input_data None --log_metrics_at_train True --gpus 0 --offline --save_checkpoints True

While the above error is terminal, there are two runtime behaviors that are potentially useful in diagnosing the problem?

../lib/python3.9/site-packages/pytorch_lightning/utilities/distributed.py:68: UserWarning: training_step returned None if it was on purpose, ignore this warning...

And also, the loss for epoch 0 is nan

I appreciate your help.

**Comment by meitarronen** [comment]
Hey @eichlerwest,
Rergarding the loss equal non in the first epoch -- it is not a bug, it is designed to be like this in the first epoch.
In that epoch we are simply gathering all the embedded data to compute its mean and covariance. 
When computing the covariance there seems to be a problem with the resulting covariance matrix. We will look into that with the arguments you sent.
Is there a specific reason you are using "None" for the transform_input_data parameter? It could be the issue. Have you tried using it with the default setting (normalize)?

**Comment by EvaFlower** [comment]
> Hey @eichlerwest,
> Rergarding the loss equal non in the first epoch -- it is not a bug, it is designed to be like this in the first epoch.
> In that epoch we are simply gathering all the embedded data to compute its mean and covariance.
> When computing the covariance there seems to be a problem with the resulting covariance matrix. We will look into that with the arguments you sent.
> Is there a specific reason you are using "None" for the transform_input_data parameter? It could be the issue. Have you tried using it with the default setting (normalize)?

@meitarronen 
Hi, I met the same issue even without "--transform_input_data None". This issue seems same as https://github.com/BGU-CS-VIL/DeepDPM/issues/4#issue-1213555285

Could you help figure it out? Thanks!

**Comment by eichlerwest** [comment]
> Hey @eichlerwest,
> Rergarding the loss equal non in the first epoch -- it is not a bug, it is designed to be like this in the first epoch.
> In that epoch we are simply gathering all the embedded data to compute its mean and covariance.
> When computing the covariance there seems to be a problem with the resulting covariance matrix. We will look into that with the arguments you sent.
> Is there a specific reason you are using "None" for the transform_input_data parameter? It could be the issue. Have you tried using it with the default setting (normalize)?

I cut & paste from the README while trying to learn how the code works - there was no other specific reason.

**Comment by meitarronen** [comment]
Can you please try with --transform_input_data normalize ?
This error means that your data's sample covariance matrix is not positive definite. Can you please try using the --transform_input_data flag and let me know how it went?

Also, it could mean that at least one of your variables can be expressed as a linear combination of the others. So it might be helpful to check that (I would try to add the features sequentially and see when it breaks)

**Comment by EvaFlower** [comment]
> Can you please try with --transform_input_data normalize ? This error means that your data's sample covariance matrix is not positive definite. Can you please try using the --transform_input_data flag and let me know how it went?
> 
> Also, it could mean that at least one of your variables can be expressed as a linear combination of the others. So it might be helpful to check that (I would try to add the features sequentially and see when it breaks)

Hi @meitarronen ,

Thanks for your follow-up reply.

I have tried with --transform_input_data normalize, but the error still occurred. 

The error information seemed referring to the matrix model_mus[k] in 
""
File "DeepDPM/src/clustering_models/clusternet_modules/utils/training_utils.py", line 235, in cluster_loss_function
    gmm_k = MultivariateNormal(model_mus[k].double().to(device=self.device), model_covs[k].double().to(device=self.device))
""
is not a positive definite.

So I tried to add the following code:
model_covs[k] = model_covs[k] + torch.eye(model_covs[k].shape[-1]) * eps,
making the model_covs[k] as a positive definite. When eps  is set to 1e-3, the error is gone. However, the obtained clustering results are bad.

Could you give some helps?

Thanks,

**Comment by wiegertj** [comment]
Hi @EvaFlower,

I ran into the very same problem. I use a vanilla convolutional AE as feature extractor. However, the latent space had too little structure as all data points were very close together. As soon as I constrained it for distance consistency in the latent space, I no longer ran into this error. 
So I assume this can also happen if your feature space is somehow cursed. Furthermore setting the prior_sigma too small also resulted in this error (setting alpha as an alternative did not).
Long story short, I would check my latent space structure using e.g. TSNE and be careful with setting prior_sigma too small.
That's my empirical experience, maybe there is some theoretical explanation for this as well.

Hope it might help someone.
